docs/specification.md

```diff
@@ -0,0 +1,127 @@
+* `addToWhitelist (address)` onlyAdmin.
+  * Description: adds address to `whitelist`.
+  * Constraints: `onlyAdmin`.
+* `removeFromWhitelist (address)` onlyAdmin.
+  * Description: removes address from `whitelist`.
+  * Constraints: `onlyAdmin`.
+* `isWhitelisted (view address bool)`
+  * Description: returns `True` if the address is whitelisted.
+* `addToBlacklist (address)` onlyAdmin.
+  * Description: adds address to `blacklist`.
+  * Constraints: `onlyAdmin`.
+* `removeFromBlacklist (address)` onlyAdmin.
+  * Description: removes address from `blacklist`.
+  * Constraints: `onlyAdmin`.
+* `isBlacklisted (view address bool)`
+  * Description: returns `True` if the address is blacklisted.
+
+Additionaly, it would be nice to have the following entrypoint:
+* `checkConstraints ([address, constraint])`.
+  * Description: here `constraint` is property that needs to be checked for given
+  address. Can be one of the following:
+  `isWhitelisted`, `isNonBlacklisted`.
+  This entrypoint doesn't do anything in case all constraints are satisfied
+  and fails with an error otherwise.
+
+This contract implements regulatory service for the **Holdings** conract.
+
+## Holdings contract
+
+Approximate storage structure:
+```haskell
+data Storage = Storage
+  { name :: MText
+  , symbol :: MText
+  , id :: MText
+  , safelistAddress :: Maybe (TAddress Safelist.Storage)
+  , owner :: Address
+  , admin :: Address
+  , ledger :: BigMap Address (Map Address Natural, Natural)
+  , paused :: Bool
+  , transferable :: Bool
+  , totalMinted :: Natural
+  , totalBurned :: Natural
+  }
+```
+
+Required `Holdings` entrypoints:
+* `setName (string :newName)`
+  * Description: updates `name` field.
+  * Constraints: `onlyAdmin` `isNotPaused`.
+* `setSymbol (string :newSymbol)`
+  * Description: updates `symbol` field.
+  * Constraints: `onlyAdmin` `isNotPaused`.
+* `setSafelistAddress ((option address) :mbNewSafelistAddress)`
+  * Description: updates `safelistAddress` field, which is used for safelist contract calls.
+  * Constraints: `onlyOwner`.
+* `setAdminAddress (address)`
+  * Description: updates `admin` field.
+  * Constraints: `onlyOwner` `isNotPaused`.
+* `isAdmin (view address bool)`
+  * Description: returns `True` if the address is admin.
+* `transfer (address :from, address :to, nat :value)`
+  * Description: transfer given amount of tokens from one address to another.
+  * Constraints: `onlyAdmin`, `isNotPaused`, `isTransferable`.
+  * Safelist constraints: `sender` is whitelisted, `sender` and `receiver` are nonblacklisted.
+* `approve (address :spender, nat :value)`
+  * Description: approve given amout of token to be spent by given address from `sender`.
+  * Constraints: `isNotPaused`.
+  * Safelist constraints: `spender` is whitelisted, `spender` and `sender` are nonblacklisted.
+* `getAllowance (view (address :owner, address :spender) natural)`
+  * Description: returns approval amount for two given addresses.
+* `view getBalance (address :owner)`
```

Comment:
It should be `getBalance (view (address :owner) nat)`. Please also update `getTotalSupply`.